assertion/function/analyzer.go

```diff
@@ -241,6 +257,175 @@ func run(pass *analysis.Pass) (result interface{}, _ error) {
+// FuncParam producer or UseAsReturn consumer or both, from the contracted functions to the callers
+// of all the contracted functions. This is necessary because we have created new
+// producers/consumers for argument pass or result return at every call site of the contracted
+// function. In order to connect such producers/consumers back to the contracted functions, we
+// create new full triggers that duplicates the original full triggers of the contracted functions
+// but uses the new producers/consumers instead.
+func duplicateFullTriggersFromContractedFunctionsToCallers(
+	pass *analysis.Pass,
+	funcContracts functioncontracts.Map,
+	funcTriggers [][]annotation.FullTrigger,
+	funcResults map[*types.Func]*functionResult,
+) {
+
+	// Find all the calls to contracted functions
+	// callsByCtrtFunc is a mapping contracted function -> caller -> call expression
+	callsByCtrtFunc := map[*types.Func]map[*types.Func]*ast.CallExpr{}
+	for funcObj, r := range funcResults {
+		for ctrFunc, calls := range findCallsToContractedFunctions(r.funcDecl, pass, funcContracts) {
+			for _, call := range calls {
+				// TODO: have to introduce v to make NilAway happy. Remove v when NilAway can
+				//  tolerate this.
+				v, ok := callsByCtrtFunc[ctrFunc]
+				if !ok {
+					v = map[*types.Func]*ast.CallExpr{}
+					callsByCtrtFunc[ctrFunc] = v
+				}
+				v[funcObj] = call
+			}
+		}
+	}
+
+	// For every contracted function, duplicate some of its full triggers (that involves param or
+	// return) into all the callers
+	dupTriggers := map[*types.Func][]annotation.FullTrigger{}
+	for ctrtFunc, calls := range callsByCtrtFunc {
+		// TODO: have to introduce r to make NilAway happy. Remove r when NilAway can tolerate
+		//  this.
+		r := funcResults[ctrtFunc]
+		if r == nil {
+			continue
+		}
+		for _, trigger := range r.triggers {
+			// If the full trigger has a FuncParam producer or a UseAsReturn consumer, then create
+			// a duplicated (possibly controlled) full trigger from it and add the created full
+			// trigger to every caller.
+			_, isParamProducer := trigger.Producer.Annotation.(annotation.FuncParam)
+			_, isReturnConsumer := trigger.Consumer.Annotation.(annotation.UseAsReturn)
+			if !isParamProducer && !isReturnConsumer {
+				// No need to duplicate the full trigger
+				continue
+			}
+			// Duplicate the full trigger in every caller
+			for caller, callExpr := range calls {
+				dupTrigger := duplicateFullTrigger(trigger, ctrtFunc, callExpr, pass,
+					isParamProducer, isReturnConsumer)
+
+				// Store the duplicated full trigger
+				dupTriggers[caller] = append(dupTriggers[caller], dupTrigger)
+			}
+		}
+	}
+
+	// Update funcTriggers with duplicated triggers
+	for funcObj, triggers := range dupTriggers {
+		// TODO: have to introduce r to make NilAway happy. Remove r when NilAway can tolerate
+		//  this.
+		r := funcResults[funcObj]
+		if r == nil {
+			continue
+		}
+		funcTriggers[r.index] = append(r.triggers, triggers...)
```

Comment:
Wouldn't this make more sense as `funcTriggers[r.index] = append(funcTriggers[r.index], triggers...)`? Just in case we ever add another feature that appends to `funcTriggers[r.index]`? (I believe line 233 has already added `r.triggers` here, unless I am not understanding the control flow)

Reply:
I agree.